h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -589,8 +592,8 @@ public void onModelFailure(Model failedModel, Consumer<Object[]> withFailedModel
           
           int[] hyperparamIndices = new int[_currentHyperParamNames.length];
           
-          do {
-            if(_hyperParamSubspaces.length != 0) {
+          for (int j=0; j < Math.max(1e4, _maxHyperSpaceSize); j++) {
```

Comment:
`max(1e4, _maxHyperSpaceSize)` is the maximum number of unsuccessful attempts to find a new parameter configuration. This can potentially cause it to finish prematurely but to exhaustively go through all possible parameter configuration we have better way (CartesianWalker).

`1e4` was chosen simply because it is high enough to make it unlikely we would not exhaust small parameter space while being low enough to finish within a second. Test I added finds a hash collision after ~874 512 iterations and takes 7s to finish.

Reply:
magic numbers should always be stored in a constant variable (possibly documented if not clear), otherwise no one will remember what this means in a few months/

Reply:
Also, I don't understand why you need this magic number now that you have the `exhausted` variable, why not:
```java
for (int j=0; j < _maxHyperSpaceSize && !exhausted; j++) {
```

Reply:
Also, as discussed, I'm convinced that the following logic to estimate total work in `GridSearch`:
```java
      while (it.hasNext()) {
        try {
          Model.Parameters parms = it.nextModelParameters();
          gridWork += (parms._nfolds > 0 ? (parms._nfolds+1/*main model*/) : 1) *parms.progressUnits();
        } catch(Throwable ex) {
          //swallow invalid combinations
        }
      }
```
should be refactored to have the iteration logic in the `HyperspaceIterator` impl: this loop is meaningless in RGS and causes the original issue.

Reply:
> Also, I don't understand why you need this magic number now that you have the `exhausted` variable, why not:
> 
> ```java
> for (int j=0; j < _maxHyperSpaceSize && !exhausted; j++) {
> ```

ok, this comment was also related with my misunderstanding below.
And `1e4` is actually an search upper bound for small hyperspaces.

Still, please extract the magic number.

Reply:
Thanks! I extracted the magic number and added some comments. I hope they'll make it more clear.

I also refactored this logic: 
```java
while (it.hasNext()) {
        try {
          Model.Parameters parms = it.nextModelParameters();
          gridWork += (parms._nfolds > 0 ? (parms._nfolds+1/*main model*/) : 1) *parms.progressUnits();
        } catch(Throwable ex) {
          //swallow invalid combinations
        }
      }
```
but not into an `Iterator` but into the `Walker` as iterator is stateful and this logic would modify the state of the iterator.